dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -0,0 +1,261 @@
+}
+
+int process_picture_decode(void *user_data, CUVIDPICPARAMS *picture_params) {
+  FramesDecoderGpu *frames_decoder = static_cast<FramesDecoderGpu*>(user_data);
+
+  // Sending empty packet will call this callback.
+  // If we want to flush the decoder, we do not need to do anything here
+  if (frames_decoder->flush_) {
+    return 0;
+  }
+
+  CUDA_CALL(cuvidDecodePicture(frames_decoder->nvdecode_state_->decoder_, picture_params));
+
+  // Process decoded frame for output
+  CUVIDPROCPARAMS videoProcessingParameters = {};
+  videoProcessingParameters.progressive_frame = !picture_params->field_pic_flag;
+  videoProcessingParameters.second_field = 1;
+  videoProcessingParameters.top_field_first = picture_params->bottom_field_flag ^ 1;
+  videoProcessingParameters.unpaired_field = 0;
+  videoProcessingParameters.output_stream = frames_decoder->stream_;
+
+  uint8_t *frame_output = nullptr;
+
+  // Take pts of the currently decoded frame
+  int current_pts = frames_decoder->piped_pts_.front();
+  frames_decoder->piped_pts_.pop();
+
+  if (current_pts == frames_decoder->CurrentFramePts()) {
+    // Currently decoded frame is actually the one we wanted
+    frames_decoder->frame_returned_ = true;
+    frame_output = frames_decoder->current_frame_output_;
+  } else {
+    // Put currently decoded frame to the buffer for later
+    auto &slot = frames_decoder->FindEmptySlot();
+    slot.pts_ = current_pts;
+    frame_output = slot.frame_.data();
+  }
+
+  if (frames_decoder->current_copy_to_output_ == false) {
+    return 1;
+  }
+
+  CUdeviceptr frame = {};
+  unsigned int pitch = 0;
+
+  CUDA_CALL(cuvidMapVideoFrame(
+    frames_decoder->nvdecode_state_->decoder_,
+    picture_params->CurrPicIdx,
+    &frame,
+    &pitch,
+    &videoProcessingParameters));
+
+  // TODO(awolant): Benchmark, if copy would be faster
+  yuv_to_rgb(
+    reinterpret_cast<uint8_t *>(frame),
+    pitch,
+    frame_output,
+    frames_decoder->Width()* 3,
+    frames_decoder->Width(),
+    frames_decoder->Height(),
+    frames_decoder->stream_);
+  CUDA_CALL(cuvidUnmapVideoFrame(frames_decoder->nvdecode_state_->decoder_, frame));
+
+  return 1;
+}
+}  // namespace detail
+
+FramesDecoderGpu::FramesDecoderGpu(const std::string &filename, cudaStream_t stream) :
+    FramesDecoder(filename),
+    frame_buffer_(num_decode_surfaces_),
+    stream_(stream) {
+    nvdecode_state_ = std::make_unique<NvDecodeState>();
+
+    const AVBitStreamFilter *bsf = av_bsf_get_by_name("h264_mp4toannexb");
+    DALI_ENFORCE(av_bsf_alloc(bsf, &bsfc_) >= 0);
+    DALI_ENFORCE(avcodec_parameters_copy(
+      bsfc_->par_in, av_state_->ctx_->streams[0]->codecpar) >= 0);
+    DALI_ENFORCE(av_bsf_init(bsfc_) >= 0);
+
+    filtered_packet_ = av_packet_alloc();
+    DALI_ENFORCE(filtered_packet_, "Could not allocate av packet");
+
+    // Create nv decoder
+    CUVIDDECODECREATEINFO decoder_info = {};
+    memset(&decoder_info, 0, sizeof(CUVIDDECODECREATEINFO));
```

Comment:
I'm not sure if you need `decoder_info = {}` when you call memset later.

Reply:
Done